src/odbc_connection.cpp

```diff
@@ -2123,43 +2126,71 @@ SQLRETURN ODBCConnection::FetchAll(QueryData *data) {
 
     // Iterate over each column, putting the data in the row object
     for (int i = 0; i < data->columnCount; i++) {
-      row[i].size = data->columns[i]->StrLen_or_IndPtr;
-      if (row[i].size == SQL_NULL_DATA || row[i].size == SQL_NO_TOTAL) {
-        row[i].char_data = NULL;
+
+      if (data->columns[i]->StrLen_or_IndPtr == SQL_NULL_DATA) {
+        row[i].size = SQL_NULL_DATA;
       } else {
         switch (data->columns[i]->bind_type) {
-          case SQL_C_DOUBLE:
-            row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
-            break;
+        case SQL_C_DOUBLE:
+          row[i].double_data = *(SQLDOUBLE*)data->boundRow[i];
+          break;
 
-          case SQL_C_UTINYINT:
-            row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
-            break;
+        case SQL_C_UTINYINT:
+          row[i].tinyint_data = *(SQLCHAR*)data->boundRow[i];
+          break;
 
-          case SQL_C_USHORT:
-            row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
-            break;
+        case SQL_C_USHORT:
+          row[i].smallint_data = *(SQLUSMALLINT*)data->boundRow[i];
+          break;
 
-          case SQL_C_SLONG:
-            row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
-            break;
+        case SQL_C_SLONG:
+          row[i].integer_data = *(SQLINTEGER*)data->boundRow[i];
+          break;
 
-          case SQL_C_UBIGINT:
-            row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
-            break;
+        case SQL_C_UBIGINT:
+          row[i].ubigint_data = *(SQLUBIGINT*)data->boundRow[i];
+          break;
 
-          case SQL_C_WCHAR:
-            row[i].wchar_data = new SQLWCHAR[row[i].size + 2]();
-            memcpy(row[i].wchar_data, data->boundRow[i], row[i].size);
-            break;
+        case SQL_C_BINARY:
+          row[i].char_data = new SQLCHAR[row[i].size]();
+          memcpy(row[i].char_data, data->boundRow[i], row[i].size);
+          break;
 
-          case SQL_C_CHAR:
-          default:
-            row[i].char_data = new SQLCHAR[row[i].size + 2]();
-            memcpy(row[i].char_data, data->boundRow[i], row[i].size);
-            break;
-        }
-        row->bind_type = data->columns[i]->bind_type;
+        case SQL_C_WCHAR:
+          row[i].size = wcslen((const wchar_t *)data->boundRow[i]);
+          row[i].wchar_data = new SQLWCHAR[(row[i].size + 1) * sizeof(SQLWCHAR)]();
```

Comment:
```suggestion
          row[i].wchar_data = new SQLWCHAR[row[i].size + 1]();
```

Array new uses number of elements in the array, not number of bytes.

Reply:
There are some more areas in the code that could be cleaned up similarly, but they will require a little more thinking, e.g.:
```
case SQL_WCHAR:
case SQL_WVARCHAR:
case SQL_WLONGVARCHAR:
  column->buffer_size = (column->ColumnSize + 1) * sizeof(SQLWCHAR);
  column->bind_type = SQL_C_WCHAR;
  data->boundRow[i] = new SQLWCHAR[column->buffer_size]();
  break;
...
data->sqlReturnCode = SQLBindCol(
      data->hSTMT,              // StatementHandle
      i + 1,                    // ColumnNumber
      column->bind_type,        // TargetType
      data->boundRow[i],        // TargetValuePtr
      column->buffer_size,      // BufferLength
      &column->StrLen_or_IndPtr // StrLen_or_Ind
    );
...
Here, buffer_size is used both to initialize array (not in bytes, twice as big as it needs to be) and to pass to `SQLBindCol`'s BufferLength (in bytes, the correct size)